contracts/modules/v2/StakingPool.sol

```diff
@@ -249,63 +284,79 @@ abstract contract StakingPool is IStakingPool, ERC721 {
-      ++lastBucket;
-      currentAllocation -= productBuckets[productId][lastBucket].allocationCut;
+    {
+      uint lastBucket = products[productId].lastBucket;
+
+      // process expirations
+      while (lastBucket < currentBucket) {
+        ++lastBucket;
+        allocatedStake -= productBuckets[productId][lastBucket].allocationCut;
+      }
     }
 
-    uint firstGroupId = block.timestamp / GROUP_SIZE;
-    // group expiration must exceed the cover period
-    uint firstUsableGroupId = (block.timestamp + period) / GROUP_SIZE;
-    uint unusableShares = 0;
+    uint availableStake;
+    {
+      // group expiration must exceed the cover period
+      uint _firstAvailableGroupId = (block.timestamp + period) / GROUP_SIZE;
+      uint _firstActiveGroupId = block.timestamp / GROUP_SIZE;
 
-    for (uint i = firstGroupId; i < firstUsableGroupId; ++i) {
-      unusableShares += stakeGroups[i].stakeShares;
-    }
+      // start with the entire supply and subtract unavailable groups
+      uint _stakeSharesSupply = stakeSharesSupply;
+      uint availableShares = _stakeSharesSupply;
 
-    uint _activeStake = activeStake;
-    uint _stakeSharesSupply = stakeSharesSupply;
-    uint usableShares = _stakeSharesSupply - unusableShares;
+      for (uint i = _firstActiveGroupId; i < _firstAvailableGroupId; ++i) {
+        availableShares -= stakeGroups[i].stakeShares;
+      }
 
-    // can be used total
-    uint usableStake = _activeStake * usableShares / _stakeSharesSupply;
-    // can be used by this product
-    usableStake = usableStake * products[productId].weight / WEIGHT_DENOMINATOR;
+      // total stake available without applying product weight
+      availableStake = activeStake * availableShares / _stakeSharesSupply;
+      // total stake available for this product
+      availableStake = availableStake * products[productId].weight / WEIGHT_DENOMINATOR;
+    }
 
     // could happen if is 100% in-use or if product weight is changed
-    if (currentAllocation >= usableStake) {
+    if (allocatedStake >= availableStake) {
+      // store expirations
+      products[productId].allocatedStake = allocatedStake;
+      products[productId].lastBucket = currentBucket;
       return (0, 0);
     }
 
-    uint maxAllocation = usableStake - currentAllocation;
-    newAllocation = min(amountInNXM, maxAllocation);
+    uint usableStake = availableStake - allocatedStake;
+    newAllocation = min(productStakeAmount, usableStake);
 
     premium = calculatePremium(
-      currentAllocation,
-      maxAllocation,
+      allocatedStake,
+      usableStake,
       newAllocation,
-      initialPriceRatio,
       period
     );
 
-    currentAllocation += newAllocation;
-    products[productId].currentAllocation = currentAllocation;
+    products[productId].allocatedStake = allocatedStake + newAllocation;
+    products[productId].lastBucket = currentBucket;
 
-    // ceil = fn(a, b) => (a + b - 1) / b
+    // divCeil = fn(a, b) => (a + b - 1) / b
```

Comment:
Move this to a pure function instead.

Reply:
Will move when it will be used in more than one place, a comment is sufficient until then.